### PR TITLE
[Preview] AI Integration: Adds diagnosticScope support and SDK Instrumentation

### PR DESCRIPTION
**Pull request by sourabh1007**
## Description

This PR includes following changes:
1. Copied Diagnostic Scope files from Azure Core
2. Use Diagnostic Scope to create Activity before running operation specific logic
3. Add Preview Tag in this feature.
4. Created parent of all responses i.e `IOpenTelemetryResponse` in order to collect request/response related data.

This PR **DOESN'T** include :
1. Setting up values in the response which can be set as open telemetry attribute.
2. Setting up Account level attributes

Since, This PR has large number of files so listing important parts:
1. `Azure.Core `files have been added here https://github.com/Azure/azure-cosmos-dotnet-v3/tree/1bd14a9946fc6ded350755214224d98621df623d/Microsoft.Azure.Cosmos/src/OSS
3.  This file decides if this feature is under preview or not 
https://github.com/Azure/azure-cosmos-dotnet-v3/blob/1bd14a9946fc6ded350755214224d98621df623d/Microsoft.Azure.Cosmos/src/Telemetry/OpenTelemetry/OpenTelemetryRecorderFactory.cs
4.  Here is the logic of capturing attributes
https://github.com/Azure/azure-cosmos-dotnet-v3/blob/1bd14a9946fc6ded350755214224d98621df623d/Microsoft.Azure.Cosmos/src/Resource/ClientContextCore.cs#L469
5. This class is parent of all the response classes and it will have all the relevant values in the future PRs.
https://github.com/Azure/azure-cosmos-dotnet-v3/blob/1bd14a9946fc6ded350755214224d98621df623d/Microsoft.Azure.Cosmos/src/Telemetry/OpenTelemetry/IOpenTelemetryResponse.cs

## Architecture Diagram
TBD

## Type of change
- [] New feature (non-breaking change which adds functionality)

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/Azure.Core/AppContextSwitchHelper.cs [hunk 1]
```diff
@@ -0,0 +1,42 @@
+﻿//------------------------------------------------------------
+// Copyright (c) Microsoft Corporation.  All rights reserved.
+//------------------------------------------------------------
+
```
Comment (FabianMeiswinkel) [inline Microsoft.Azure.Cosmos/src/Azure.Core/AppContextSwitchHelper.cs:4]: Does it make sense to add info here form where the file was copied?

File: Microsoft.Azure.Cosmos/src/Telemetry/Diagnostics/CosmosInstrumentation.cs [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+                }
+                this.scope.AddAttribute(attributeKey, attributeValue);
+            }
+        }
+
+        public void Record(ITrace trace)
+        {
+            if (this.scope.IsEnabled)
+            {
+                CosmosTraceDiagnostics diagnostics = new CosmosTraceDiagnostics(trace);
```
Comment (FabianMeiswinkel) [inline Microsoft.Azure.Cosmos/src/Telemetry/Diagnostics/CosmosInstrumentation.cs:47]: Would this result in creating another CosmosTraceDiagnostics instance on another part of the request pipeline for the same ITrace instance - if so we should reuse the created instance?

File: Microsoft.Azure.Cosmos/src/Telemetry/Diagnostics/Filters/DiagnosticsFilterHelper.cs [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+        /// <summary>
+        /// Allow only when either of below is True
+        /// 1) Latency is not more than 100 ms
+        /// 2) Request Charge is more than 100 RUs
+        /// 3) HTTP status code is not Success
+        /// 
+        /// </summary>
+        /// <returns>true or false</returns>
+        public static bool IsAllowed(
```
Comment (FabianMeiswinkel) [inline Microsoft.Azure.Cosmos/src/Telemetry/Diagnostics/Filters/DiagnosticsFilterHelper.cs:23]: This is used to determine whether we collect Client Telemetry - not request diagnostics right now - that is unintentional, isn't it?

File: Microsoft.Azure.Cosmos/src/Telemetry/Diagnostics/CosmosInstrumentation.cs (hunk 2 above)
Comment (sourabh1007) [reply Microsoft.Azure.Cosmos/src/Telemetry/Diagnostics/CosmosInstrumentation.cs:47]: It will be called only in case of disposing of root instance of ITrace. So, this instance will be created only once.

https://github.com/Azure/azure-cosmos-dotnet-v3/blob/befde9e43ac1afe18326bdd37d2beb65be803d17/Microsoft.Azure.Cosmos/src/Tracing/Trace.cs#L100

File: Microsoft.Azure.Cosmos/src/Telemetry/Diagnostics/CosmosInstrumentation.cs [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+                this.Record(OTelAttributes.Region, ClientTelemetryHelper.GetContactedRegions(diagnostics));
+
+                this.Record(OTelAttributes.RequestDiagnostics, diagnostics.ToString());
```
[the remark this answers is not in the document]
Comment (sourabh1007) [reply Microsoft.Azure.Cosmos/src/Telemetry/Diagnostics/CosmosInstrumentation.cs:51]: Filter logic will be added as part of other PR. In this PR, I want to make sure that We have all the required files and this feature is enabled only in preview mode.

File: Microsoft.Azure.Cosmos/src/Telemetry/Diagnostics/Filters/DiagnosticsFilterHelper.cs (hunk 3 above)
Comment (sourabh1007) [reply Microsoft.Azure.Cosmos/src/Telemetry/Diagnostics/Filters/DiagnosticsFilterHelper.cs:23]: It is not related to client telemetry in any way...it is different filter class written only for AI integration feature.

**Review by FabianMeiswinkel — APPROVED** [review]
LGTM - Thanks

File: Microsoft.Azure.Cosmos/src/Azure.Core/AppContextSwitchHelper.cs [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+#nullable enable
+
+namespace Azure.Core
+{
+    using System;
+    /// <summary>
+    /// Helper for interacting with AppConfig settings and their related Environment variable settings.
+    /// </summary>
+    internal static class AppContextSwitchHelper
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Azure.Core/AppContextSwitchHelper.cs:15]: Please move the files into the OSS folder. https://github.com/Azure/azure-cosmos-dotnet-v3/tree/master/Microsoft.Azure.Cosmos/src/OSS

File: Microsoft.Azure.Cosmos/src/Telemetry/Diagnostics/OTelAttributes.cs [hunk 6]
```diff
@@ -0,0 +1,36 @@
+﻿//------------------------------------------------------------
+// Copyright (c) Microsoft Corporation.  All rights reserved.
+//------------------------------------------------------------
+
+namespace Microsoft.Azure.Cosmos.Telemetry.Diagnostics
+{
+    internal class OTelAttributes
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Telemetry/Diagnostics/OTelAttributes.cs:7]: Why is this abbreviated? Use OpenTelemetryAttributes

File: Microsoft.Azure.Cosmos/src/Handler/RequestInvokerHandler.cs [hunk 7]
```diff
@@ -122,6 +123,7 @@ internal class RequestInvokerHandler : RequestHandler
             ITrace trace,
             CancellationToken cancellationToken)
         {
+            ResponseMessage response = null;
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Handler/RequestInvokerHandler.cs:126]: nit: inline variable declaration for readability. No reason for it to be declare this high up if it always just returns in the scope of the one catch block.

File: Microsoft.Azure.Cosmos/src/Handler/RequestInvokerHandler.cs [hunk 8]
```diff
@@ -206,7 +216,11 @@ internal class RequestInvokerHandler : RequestHandler
                             }
                             catch (CosmosException ex)
                             {
-                                return ex.ToCosmosResponseMessage(request);
+                                response = ex.ToCosmosResponseMessage(request);
+
+                                childTrace.CosmosInstrumentation.MarkFailed(ex);
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Handler/RequestInvokerHandler.cs:221]: Why isn't this being done higher up in the ClientContextCore?

File: Microsoft.Azure.Cosmos/src/Resource/CosmosExceptions/CosmosObjectDisposedException.cs [hunk 9]
```diff
@@ -44,6 +44,8 @@ internal class CosmosObjectDisposedException : ObjectDisposedException
                 throw new ArgumentNullException(nameof(trace));
             }
 
+            trace.CosmosInstrumentation.MarkFailed(originalException);
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Resource/CosmosExceptions/CosmosObjectDisposedException.cs:47]: Shouldn't this log the new exception with the diagnostics?

File: Microsoft.Azure.Cosmos/src/Telemetry/Diagnostics/CosmosInstrumentation.cs [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+
+namespace Microsoft.Azure.Cosmos.Telemetry.Diagnostics
+{
+    using System;
+    using System.Net;
+    using global::Azure.Core.Pipeline;
+    using Microsoft.Azure.Cosmos.Diagnostics;
+    using Microsoft.Azure.Cosmos.Tracing;
+
+    internal class CosmosInstrumentation : ICosmosInstrumentation
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Telemetry/Diagnostics/CosmosInstrumentation.cs:13]: Why is this called ICosmosInstrumentation?  Shouldn't it be more like CosmosOpenTelemetry or something with more details on what it actually is.

File: Microsoft.Azure.Cosmos/src/Telemetry/Diagnostics/Filters/DiagnosticsFilterHelper.cs [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+        /// 3) HTTP status code is not Success<br></br>
+        /// </summary>
+        /// <returns>true or false</returns>
+        public static bool IsAllowed(
+            TimeSpan latency, 
+            double requestcharge, 
+            HttpStatusCode statuscode)
+        {
+            return latency > TimeSpan.FromMilliseconds(DiagnosticsFilterHelper.latencyThresholdInMs) ||
+                   requestcharge > DiagnosticsFilterHelper.requestChargeThresholdInRu ||
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Telemetry/Diagnostics/Filters/DiagnosticsFilterHelper.cs:28]: Why is this always logging based on a RU charge? Batch/bulk/query can easily cause this to log everything.

File: Microsoft.Azure.Cosmos/src/Telemetry/Diagnostics/Filters/DiagnosticsFilterHelper.cs [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+                   !statuscode.IsSuccess();
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Telemetry/Diagnostics/Filters/DiagnosticsFilterHelper.cs:29]: We should probably exclude 404.

File: Microsoft.Azure.Cosmos/src/Tracing/ITrace.cs [hunk 13]
```diff
@@ -120,5 +126,6 @@ interface ITrace : IDisposable
         /// </summary>
         /// <param name="traceDatum"></param>
         void UpdateRegionContacted(TraceDatum traceDatum);
+
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Tracing/ITrace.cs:129]: nit: remove extra line

File: Microsoft.Azure.Cosmos/src/Telemetry/Diagnostics/CosmosInstrumentation.cs [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+                }
+                if (attributeKey.Equals(OTelAttributes.StatusCode))
+                {
+                    this.statusCode = (HttpStatusCode)attributeValue;
+                }
+                this.scope.AddAttribute(attributeKey, attributeValue);
+            }
+        }
+
+        public void Record(ITrace trace)
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Telemetry/Diagnostics/CosmosInstrumentation.cs:43]: Can you explain the ownership? This seems rather confusing that open telemetry uses ITrace, but ITrace also has a reference to it. 

Would a better model be to have ITrace call the open telemetry APIs instead of having this dual ownership?

File: Microsoft.Azure.Cosmos/src/Handler/RequestInvokerHandler.cs [hunk 15]
```diff
@@ -285,9 +307,13 @@ internal class RequestInvokerHandler : RequestHandler
                         request.ContainerId = cosmosContainerCore?.Id;
                         request.DatabaseId = cosmosContainerCore?.Database.Id;
                     }
+
                     requestEnricher?.Invoke(request);
 
-                    return await this.SendAsync(request, cancellationToken);
+                    response = await this.SendAsync(request, cancellationToken);
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Handler/RequestInvokerHandler.cs:313]: nit: Why not just return this statement?

File: Microsoft.Azure.Cosmos/src/Telemetry/Diagnostics/CosmosInstrumentation.cs [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+            this.scope = scope;
+
+            this.scope.Start();
+        }
+
+        public void Record(string attributeKey, object attributeValue)
+        {
+            if (this.scope.IsEnabled)
+            {
+                if (attributeKey.Equals(OTelAttributes.RequestCharge))
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Telemetry/Diagnostics/CosmosInstrumentation.cs:31]: String comparison should use the StringComparison.OrdinalIgnoreCase for performance?

File: Microsoft.Azure.Cosmos/src/Telemetry/Diagnostics/CosmosInstrumentation.cs [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+
+        public void MarkFailed(Exception exception)
+        {
+            if (this.scope.IsEnabled)
+            {
+                this.scope.AddAttribute(OTelAttributes.ExceptionMessage, exception.Message);
+                this.scope.AddAttribute(OTelAttributes.ExceptionStacktrace, exception.StackTrace);
+                this.scope.AddAttribute(OTelAttributes.ExceptionType, exception.GetType());
+
+                this.scope.Failed(exception);
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Telemetry/Diagnostics/CosmosInstrumentation.cs:69]: nit: If we are logging the exception, do we need the other promoted properties like the StackTrace?

File: Microsoft.Azure.Cosmos/src/Telemetry/Diagnostics/Filters/DiagnosticsFilterHelper.cs [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+        /// 2) Request Charge is more than 100 RUs<br></br>
+        /// 3) HTTP status code is not Success<br></br>
+        /// </summary>
+        /// <returns>true or false</returns>
+        public static bool IsAllowed(
+            TimeSpan latency, 
+            double requestcharge, 
+            HttpStatusCode statuscode)
+        {
+            return latency > TimeSpan.FromMilliseconds(DiagnosticsFilterHelper.latencyThresholdInMs) ||
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Telemetry/Diagnostics/Filters/DiagnosticsFilterHelper.cs:27]: nit: Instead of having the latencyThresholdInMs as the const, have the TimeSpan as a const/readonly

File: Microsoft.Azure.Cosmos/src/Telemetry/Diagnostics/ICosmosInstrumentation.cs [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+
+    /// <summary>
+    /// Cosmos Instrumentation Interface
+    /// </summary>
+#if INTERNAL
+    public
+#else
+    internal
+#endif 
+        interface ICosmosInstrumentation : IDisposable
```
[the remark this answers is not in the document]
Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/Telemetry/Diagnostics/ICosmosInstrumentation.cs:18]: While we can use interfaces internally, public contracts we tend to use abstract classes so in the case of having to add new properties or methods, its not a breaking change

File: Microsoft.Azure.Cosmos/src/Handler/RequestInvokerHandler.cs (hunk 8 above)
Comment (sourabh1007) [reply Microsoft.Azure.Cosmos/src/Handler/RequestInvokerHandler.cs:221]: This is the final place where exception will come. So, it is more convenient to catch an exception and I will get enriched exception object here.

File: Microsoft.Azure.Cosmos/src/Telemetry/Diagnostics/Filters/DiagnosticsFilterHelper.cs (hunk 11 above)
Comment (sourabh1007) [reply Microsoft.Azure.Cosmos/src/Telemetry/Diagnostics/Filters/DiagnosticsFilterHelper.cs:28]: yes it will,.. @FabianMeiswinkel  do you have any suggestion to handle this?

File: Microsoft.Azure.Cosmos/src/Telemetry/Diagnostics/Filters/DiagnosticsFilterHelper.cs (hunk 12 above)
Comment (FabianMeiswinkel) [reply Microsoft.Azure.Cosmos/src/Telemetry/Diagnostics/Filters/DiagnosticsFilterHelper.cs:29]: +1 - at least 404 with sub status code 0

Comment (FabianMeiswinkel) [reply Microsoft.Azure.Cosmos/src/Telemetry/Diagnostics/Filters/DiagnosticsFilterHelper.cs:29]: same is true for 409 and 412

Comment (FabianMeiswinkel) [reply Microsoft.Azure.Cosmos/src/Telemetry/Diagnostics/Filters/DiagnosticsFilterHelper.cs:29]: and probably 403 and 401

Comment (sourabh1007) [reply Microsoft.Azure.Cosmos/src/Telemetry/Diagnostics/Filters/DiagnosticsFilterHelper.cs:29]: I was wondering..if diagnostics would be helpful, if cx is getting unexpected 404 errors.

File: Microsoft.Azure.Cosmos/src/Telemetry/Diagnostics/Filters/DiagnosticsFilterHelper.cs (hunk 11 above)
Comment (FabianMeiswinkel) [reply Microsoft.Azure.Cosmos/src/Telemetry/Diagnostics/Filters/DiagnosticsFilterHelper.cs:28]: I think with the new model we have not riggering diagnostcis on request charge is ok - as long as we have the activityId/CorrelatedActivityId in the event - any query investigation into RU outliers can happen server-side then

File: Microsoft.Azure.Cosmos/src/Telemetry/Diagnostics/CosmosInstrumentation.cs (hunk 14 above)
Comment (sourabh1007) [reply Microsoft.Azure.Cosmos/src/Telemetry/Diagnostics/CosmosInstrumentation.cs:43]: OpenTelemetry is part of Trace....but we need diagnostics in Open Telemetry which is part of Trace. I have made the changes to indicate the ownership clearly.

File: Microsoft.Azure.Cosmos/src/Telemetry/Diagnostics/CosmosInstrumentation.cs (hunk 17 above)
Comment (sourabh1007) [reply Microsoft.Azure.Cosmos/src/Telemetry/Diagnostics/CosmosInstrumentation.cs:69]: These attributes will be visible on appinsights So these are required.

File: Microsoft.Azure.Cosmos/src/Telemetry/Diagnostics/Filters/DiagnosticsFilterHelper.cs (hunk 11 above)
Comment (sourabh1007) [reply Microsoft.Azure.Cosmos/src/Telemetry/Diagnostics/Filters/DiagnosticsFilterHelper.cs:28]: Removed request charge check as of now.

File: Microsoft.Azure.Cosmos/src/Handler/RequestInvokerHandler.cs [hunk 20]
```diff
@@ -177,11 +178,19 @@ internal class RequestInvokerHandler : RequestHandler
                                 }
                                 catch (DocumentClientException dce)
                                 {
-                                    return dce.ToCosmosResponseMessage(request);
+                                    ResponseMessage response = dce.ToCosmosResponseMessage(request);
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Handler/RequestInvokerHandler.cs:181]: nit: Wouldn't this simplify?
```
childTrace.OpenTelemetryAttributeRecorder.MarkFailed(dce);
return dce.ToCosmosResponseMessage(request);
```

File: Microsoft.Azure.Cosmos/src/Handler/RequestInvokerHandler.cs [hunk 21]
```diff
@@ -177,11 +178,19 @@ internal class RequestInvokerHandler : RequestHandler
...
+
+                                    childTrace.OpenTelemetryAttributeRecorder.MarkFailed(dce);
+
+                                    return response;
                                 }
                                 catch (CosmosException ce)
                                 {
-                                    return ce.ToCosmosResponseMessage(request);
+                                    ResponseMessage response = ce.ToCosmosResponseMessage(request);
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Handler/RequestInvokerHandler.cs:189]: nit: Wouldn't this simplify?
```
childTrace.OpenTelemetryAttributeRecorder.MarkFailed(ce);
return ce.ToCosmosResponseMessage(request);
```

File: Microsoft.Azure.Cosmos/src/Handler/RequestInvokerHandler.cs [hunk 22]
```diff
@@ -206,7 +215,11 @@ internal class RequestInvokerHandler : RequestHandler
                             }
                             catch (CosmosException ex)
                             {
-                                return ex.ToCosmosResponseMessage(request);
+                                ResponseMessage response = ex.ToCosmosResponseMessage(request);
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Handler/RequestInvokerHandler.cs:218]: Same potential simplification here

File: Microsoft.Azure.Cosmos/src/Handler/RequestInvokerHandler.cs [hunk 23]
```diff
@@ -223,7 +236,11 @@ internal class RequestInvokerHandler : RequestHandler
                                     subStatusCode: default,
                                     activityId: Guid.Empty.ToString(),
                                     requestCharge: default);
-                                return notFound.ToCosmosResponseMessage(request);
+                                ResponseMessage response = notFound.ToCosmosResponseMessage(request);
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Handler/RequestInvokerHandler.cs:239]: Similar simplification here:

```
childTrace.OpenTelemetryAttributeRecorder.MarkFailed(notFound);
return notFound.ToCosmosResponseMessage(request);
```

File: Microsoft.Azure.Cosmos/src/Handler/RequestInvokerHandler.cs [hunk 24]
```diff
@@ -239,7 +256,11 @@ internal class RequestInvokerHandler : RequestHandler
                                     activityId: Guid.NewGuid().ToString(),
                                     requestCharge: default);
 
-                                return goneException.ToCosmosResponseMessage(request);
+                                ResponseMessage response = goneException.ToCosmosResponseMessage(request);
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Handler/RequestInvokerHandler.cs:259]: Same potential simplification here

**Review by ealsur — APPROVED** [review]
LGTM except some potential code simplifications

File: Microsoft.Azure.Cosmos/src/Handler/RequestInvokerHandler.cs (hunk 8 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Handler/RequestInvokerHandler.cs]: The final place is the ClientContextCore where the Trace is created.  There are different cache that get created above the RequestInvokerHandler file.

File: Microsoft.Azure.Cosmos/src/Telemetry/OpenTelemetry/IRecorder.cs [hunk 25]
```diff
@@ -0,0 +1,44 @@
...
+
+    /// <summary>
+    /// Cosmos Instrumentation Interface
+    /// </summary>
+#if INTERNAL
+    public
+#else
+    internal
+#endif 
+        abstract class IRecorder : IDisposable
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Telemetry/OpenTelemetry/IRecorder.cs:18]: Why have a separate contract for this instead of just adding the methods to the ITrace contract?

File: Microsoft.Azure.Cosmos/src/Handler/RequestInvokerHandler.cs (hunk 8 above)
Comment (sourabh1007) [reply Microsoft.Azure.Cosmos/src/Handler/RequestInvokerHandler.cs]: Here we know that request were failed with some exception hence marking that activity as failure and Adding exception detail. If I try to do the same thing in `CosmosContextCore`, I need to check each response for the reason of failure and then mark that activity as failure. By marking an Activity as failure in `RequestInvokerHandler`, I have avoided that check

File: Microsoft.Azure.Cosmos/src/Tracing/Trace.cs [hunk 26]
```diff
@@ -92,6 +95,13 @@ public void UpdateRegionContacted(TraceDatum traceDatum)
         public void Dispose()
         {
             this.stopwatch.Stop();
+
+            if (this.Parent == null)
+            { 
+                this.OpenTelemetryAttributeRecorder.Record(new CosmosTraceDiagnostics(this));
```
[the remark this answers is not in the document]
Comment (sourabh1007) [reply Microsoft.Azure.Cosmos/src/Tracing/Trace.cs]: Going with new design where Open Telemetry won't be part of Traces

File: Microsoft.Azure.Cosmos/src/Telemetry/OpenTelemetry/IRecorder.cs (hunk 25 above)
Comment (sourabh1007) [reply Microsoft.Azure.Cosmos/src/Telemetry/OpenTelemetry/IRecorder.cs]: -Going with new design where Open Telemetry won't be part of Traces

File: Microsoft.Azure.Cosmos/src/Batch/TransactionalBatchResponse.cs [hunk 27]
```diff
@@ -21,7 +21,7 @@ namespace Microsoft.Azure.Cosmos
     /// Response of a <see cref="TransactionalBatch"/> request.
     /// </summary>
 #pragma warning disable CA1710 // Identifiers should have correct suffix
-    public class TransactionalBatchResponse : IReadOnlyList<TransactionalBatchOperationResult>, IDisposable
+    public class TransactionalBatchResponse : IReadOnlyList<TransactionalBatchOperationResult>, IDisposable, IOpenTelemetryResponse
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Batch/TransactionalBatchResponse.cs:24]: Addition of properties are treated as  minor.
But extending different type I guess might be treated as breaking change. 

@j82w thoughts?

File: Microsoft.Azure.Cosmos/src/Resource/Container/ContainerCore.cs [hunk 28]
```diff
@@ -118,12 +118,11 @@ internal abstract partial class ContainerCore : ContainerInternal
             return this.ClientContext.ResponseFactory.CreateContainerResponse(this, response);
         }
 
-        public async Task<int?> ReadThroughputAsync(
+        public async Task<ThroughputResponse> ReadThroughputAsync(
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Resource/Container/ContainerCore.cs:121]: @j82w will it impact compute?

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/Tracing/TraceJoiner.cs [hunk 29]
```diff
@@ -65,6 +66,8 @@ public TraceForest(IReadOnlyList<ITrace> children)
 
             public IReadOnlyList<(string, Uri)> RegionsContacted => new List<(string, Uri)>();
 
+            public IOpenTelemetryRecorder OpenTelemetryAttributeRecorder => new OpenTelemetryRecorderNoOp();
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/Tracing/TraceJoiner.cs:69]: Why is this neede?

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/Tracing/TraceWriterBaselineTests.cs [hunk 30]
```diff
@@ -35,6 +35,7 @@ namespace Microsoft.Azure.Cosmos.Tests.Tracing
     using Microsoft.Azure.Documents;
     using Microsoft.VisualStudio.TestTools.UnitTesting;
     using Newtonsoft.Json.Linq;
+    using Telemetry.Diagnostics;
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/Tracing/TraceWriterBaselineTests.cs:38]: revert white space changes

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/Tracing/EndToEndTraceWriterBaselineTests.cs [hunk 31]
```diff
@@ -1391,6 +1392,8 @@ public ITrace StartChild(string name)
                 return this.StartChild(name, TraceComponent.Unknown, TraceLevel.Info);
             }
 
+            public IOpenTelemetryRecorder OpenTelemetryAttributeRecorder { get; }
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/Tracing/EndToEndTraceWriterBaselineTests.cs:1395]: Why is this needed?

File: Microsoft.Azure.Cosmos/src/Tracing/ITrace.cs [hunk 32]
```diff
@@ -7,6 +7,7 @@ namespace Microsoft.Azure.Cosmos.Tracing
     using System;
     using System.Collections.Generic;
     using System.Runtime.CompilerServices;
+    using Telemetry.Diagnostics;
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Tracing/ITrace.cs:10]: nit: revert all the changes on files that have no code changes like this one.

File: Microsoft.Azure.Cosmos/src/Resource/Response.cs [hunk 33]
```diff
@@ -69,5 +70,25 @@ public abstract class Response<T>
         /// Gets the cosmos diagnostics information for the current request to Azure Cosmos DB service
         /// </summary>
         public abstract CosmosDiagnostics Diagnostics { get; }
+
+        /// <summary>
+        /// RequestContentLength
+        /// </summary>
+        public abstract long? RequestContentLength { get; }
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Resource/Response.cs:77]: This is a breaking change. These need to be internal virtual instead of abstract.

File: Microsoft.Azure.Cosmos/src/Resource/Container/ContainerCore.cs (hunk 28 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Resource/Container/ContainerCore.cs:121]: It will possible impact compute. Ideally they should be targeting ContainerInternal

File: Microsoft.Azure.Cosmos/src/Handler/ResponseMessage.cs [hunk 34]
```diff
@@ -169,6 +169,23 @@ public virtual CosmosDiagnostics Diagnostics
         /// </summary>
         public virtual bool IsSuccessStatusCode => this.StatusCode.IsSuccess();
 
+        /// <summary>
+        /// RequestCharge
+        /// </summary>
+        public double RequestCharge => this.Headers.RequestCharge;
+
+        /// <inheritdoc/>
+        public long? RequestContentLength => throw new NotImplementedException();
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Handler/ResponseMessage.cs:178]: Must be internal to avoid public API changes

File: Microsoft.Azure.Cosmos/src/Batch/TransactionalBatchResponse.cs (hunk 27 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Batch/TransactionalBatchResponse.cs:24]: Does IOpenTelemetryResponse show up in the public contract changes? Not sure if internal interfaces will impact public API.

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/Tracing/TraceWriterBaselineTests.cs (hunk 30 above)
Comment (sourabh1007) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/Tracing/TraceWriterBaselineTests.cs:38]: reverted

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/Tracing/TraceJoiner.cs (hunk 29 above)
Comment (sourabh1007) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/Tracing/TraceJoiner.cs:69]: reverted

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/Tracing/EndToEndTraceWriterBaselineTests.cs (hunk 31 above)
Comment (sourabh1007) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/Tracing/EndToEndTraceWriterBaselineTests.cs:1395]: reverted

File: Microsoft.Azure.Cosmos/src/Tracing/ITrace.cs (hunk 32 above)
Comment (sourabh1007) [reply Microsoft.Azure.Cosmos/src/Tracing/ITrace.cs:10]: reverted

File: Microsoft.Azure.Cosmos/src/Batch/TransactionalBatchResponse.cs (hunk 27 above)
Comment (sourabh1007) [reply Microsoft.Azure.Cosmos/src/Batch/TransactionalBatchResponse.cs:24]: 1) I didn't have to make changes in the existing test cases so it should verify that we are not changing the public contract but we are adding things in public contract.
2) Here we are adding a new Parent, so I don't think it will break the existing code but `IOpenTelemetryResponse` will be the part of public contract.

@kirankumarkolli can you please explain a scenario, where it would be a breaking change.

File: Microsoft.Azure.Cosmos/src/CosmosClient.cs [hunk 35]
```diff
@@ -500,12 +500,12 @@ protected CosmosClient()
...
         /// </returns>
-        public virtual Task<AccountProperties> ReadAccountAsync()
+        public async virtual Task<AccountProperties> ReadAccountAsync()
         {
-            return this.ClientContext.OperationHelperAsync(
+            return (await this.ClientContext.OperationHelperAsync(
                 nameof(ReadAccountAsync),
                 null,
-                (trace) => ((IDocumentClientInternal)this.DocumentClient).GetDatabaseAccountInternalAsync(this.Endpoint));
+                (trace) => ((IDocumentClientInternal)this.DocumentClient).GetDatabaseAccountResponseInternalAsync(this.Endpoint))).Resource;
```
[the remark this answers is not in the document]
Comment (sourabh1007) [reply Microsoft.Azure.Cosmos/src/CosmosClient.cs:508]: because `clientcontextcore` can return only `IOpenTelemetryResponse` while running any function.

File: Microsoft.Azure.Cosmos/src/Batch/TransactionalBatchResponse.cs (hunk 27 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Batch/TransactionalBatchResponse.cs:24]: The gates are failing and one of the failures is the contract test. Until the gates pass it's not clear what the contract changes are.

**Comment by sourabh1007** [comment]
new PR has been created with new design changes..discarding this PR.
